src/main/java/org/opensearch/security/user/User.java

```diff
@@ -67,6 +67,8 @@ public class User implements Serializable, Writeable, CustomAttributesAware {
     private String requestedTenant;
     private Map<String, String> attributes = Collections.synchronizedMap(new HashMap<>());
     private boolean isInjected = false;
+    private boolean isInternal;
+    private String authDomain;
```

Comment:
The isInternal field seems redundant given the presence of authDomain. This field essentially differentiates one specific authDomain from others.  If we included this field, it would suggest that in future PRs we might add a field `isSaml` - not be ideal.

If authDomain does not have a finite, predetermined list of values and instead denotes a specific instance of a domain, we should consider making it opaque. Alternatively consider using an object that carries both an identifier used for referencing a specific instance and a value representing the implementation of the domain in the form of an enum or class name.

Reply:
You are right that they are linked together and it's (mostly) possible to derive `isInternal` from the auth domain. 

The code for deriving `isInternal` from `authDomain` would look something like: 

```
public boolean isInternal(String authDomain) {
    boolean isInternal = false;
    // dcm is the DynamicConfigModel which can be subscribed to for changes
    restAuthDomains = Collections.unmodifiableSortedSet(dcm.getRestAuthDomains());
    for (AuthDomain ad : restAuthDomains) {
        if (ad.getName().equals(authDomain) {
            if (ad.getAuthentationBackend() instance of InternalAuthenticationBackend) {
                isInternal = true;
                break;
           }
        }
    }
    return isInternal;
}
```

I say mostly because of the problem you mentioned earlier about the fact that auth domain names are subject to change though its not likely.

Reply:
IMO I think it makes sense to have `isInternal` because of the problem where the name of the authDomain is subject to change. It's also possible to have 2 authenticators use the internal authentication backend so when thinking about equality for internal users it may not care which auth domain authenticated the user, but for other authenticators equality of the user would mean matching the authenticator that authenticated the user.

Reply:
Why don't we lock the name of the internal IdP and then we can just use a not matching name as indicative of being external. I don't know why we would need to configure the name of the internal IdP and I also don't know why we would need multiple internal IdP backed providers. 

Is there a common use case that would require multiple internal IdPs to be running at the same time?

Reply:
Auth Domain != Auth Backend

Consider the on-behalf-of auth domain where a token can be issued on behalf of a user after successfully authenticating with basic auth. When they use the on-behalf-of token they use a different auth domain to authenticate, but its the same user.

Reply:
Since this isn't used end-to-end it's hard to determine how useful/not useful the property will be.

If there are functions acting on users that should behave differently depending on what the auth domain is, we should have the auth domains handle this with polymorphism, not put the burden on the caller.

Reply:
@peternied I intend to use the changes this PR introduces in this block of the draft PR in security for scheduled job identity management: https://github.com/opensearch-project/security/pull/2773/files#diff-c56e4aaf50cd20ad928dbb7dc5026e5b2ab1fcb92396290ff4a880203dc220d1R60-R76

Inside the `IdentityPlugin.ScheduledJobIdentityManager.saveUserDetails()` call it needs to have access to a flag to indicate whether the user is internal to block the association of a scheduled job to an external user which would not be initially supported since there is not currently a mechanism put into place to resolve a user's authz ad hoc.

To address the polymorphism comment, do you think on a high level it can be split into `InternalUser` vs `ExternalUser`?

Reply:
I don't see why we need it to be domain specific though. Why would we care about the domain when we are only interested in the backend? I thought the idea is not to provide federation with these flags, just determine if they are from the internal backend or an external provider. If you have set up a SAML system and OIDC system to use the same backend that is your job to resolve the naming conflicts IMO. 

Is there something that requires the domain-specific vs. backend specific? Is it not feasible to have a backend specific flag?

Reply:
@scrawfor99 There are only 3 types of Authentication Backends in the security plugin. 1. internal, 2. LDAP, 3. NoOp. You can have a number of _authenticators_ through many means: 1. Basic Auth, 2. JWT, 3. SAML, 4, OIDC, 5. Client Cert, 6. Proxy, ...

You can technically have multiple authenticators of the same type setup simultaneously. You can definitely have multiple JWT authenticators, possibly multiple SAML authenticators. It may not represent a real production scenario, but it can happen.

The `authDomain` is the key of the authenticator in the list of `authc:` backends which ties the user back to the authenticator the user authenticated with. `isInternal` is a flag that indicates if it is an internal user or not. There is only one internal user list, so you can differentiate users as either internal or not. Maybe there is a gap of understanding. What would be the best way to address the gap? 

This approach may not be ideal and I'm open to ideas on how to change that, but if there needs to be a mechanism that only allows internal users to created scheduled jobs for extensions and not others then there needs to be a way to disambiguate the users. The alternative is to operate under the assumption that if a username can be found in the list of internal users then it is definitely an internal user. I can setup a cluster where that assumption is false. The other alternative is if its not tied to the user object then I can pass that info in the threadcontext, but that's also not a great solution.